lib/pdcleaner/pdcleaner.go

```diff
@@ -0,0 +1,249 @@
+}
+
+func NewPieceDirectoryCleaner(provAddr address.Address, cfg *config.Boost) func(lc fx.Lifecycle, dealsDB *db.DealsDB, directDealsDB *db.DirectDealsDB, legacyDeals legacy.LegacyDealManager, pd *piecedirectory.PieceDirectory, full v1api.FullNode) PieceDirectoryCleanup {
+	return func(lc fx.Lifecycle, dealsDB *db.DealsDB, directDealsDB *db.DirectDealsDB, legacyDeals legacy.LegacyDealManager, pd *piecedirectory.PieceDirectory, full v1api.FullNode) PieceDirectoryCleanup {
+
+		pdc := newPDC(dealsDB, directDealsDB, legacyDeals, pd, full, provAddr)
+
+		ctx, cancel := context.WithCancel(context.Background())
+
+		lc.Append(fx.Hook{
+			OnStart: func(_ context.Context) error {
+				mid, err := address.NewFromString(cfg.Wallets.Miner)
+				if err != nil {
+					return fmt.Errorf("failed to parse the miner ID %s: %w", cfg.Wallets.Miner, err)
+				}
+				pdc.miner = mid
+				go pdc.Start(ctx)
+				return nil
+			},
+			OnStop: func(_ context.Context) error {
+				cancel()
+				return nil
+			},
+		})
+
+		return pdc
+
+	}
+}
+
+func newPDC(dealsDB *db.DealsDB, directDealsDB *db.DirectDealsDB, legacyDeals legacy.LegacyDealManager, pd *piecedirectory.PieceDirectory, full v1api.FullNode, provAddr address.Address) *pdcleaner {
+	return &pdcleaner{
+		dealsDB:       dealsDB,
+		directDealsDB: directDealsDB,
+		legacyDeals:   legacyDeals,
+		pd:            pd,
+		full:          full,
+		miner:         provAddr,
+	}
+}
+
+func (p *pdcleaner) Start(ctx context.Context) {
+	p.ctx = ctx
+	go p.clean()
+}
+
+func (p *pdcleaner) clean() {
+	// Create a ticker with an hour tick
+	ticker := time.NewTicker(time.Hour * 6)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Infof("Starting LID clean up")
+			err := p.CleanOnce()
+			if err != nil {
+				log.Errorf("Failed to cleanup LID: %s", err)
+				continue
+			}
+			log.Info("Finished cleaning up LID")
+		case <-p.ctx.Done():
+			return
+		}
+	}
+}
+
+// CleanOnce generates a list of all Boost, Legacy and Direct deals. It then attempts to clean up the expired/slashed deals.
+// It also generated a list of all pieces in LID and tries to find any pieceMetadata with no deals in Boost, Direct or Legacy DB.
+// If such a deal is found, it is cleaned up as well
+func (p *pdcleaner) CleanOnce() error {
+	head, err := p.full.ChainHead(p.ctx)
+	if err != nil {
+		return fmt.Errorf("getting chain head: %w", err)
+	}
+	tskey := head.Key()
+	deals, err := p.full.StateMarketDeals(p.ctx, tskey)
+	if err != nil {
+		return fmt.Errorf("getting market deals: %w", err)
+	}
+
+	var boostDeals []*types.ProviderDealState
+
+	boostCompleteDeals, err := p.dealsDB.ListCompleted(p.ctx)
+	if err != nil {
+		return fmt.Errorf("getting complete boost deals: %w", err)
+	}
+	boostActiveDeals, err := p.dealsDB.ListActive(p.ctx)
+	if err != nil {
+		return fmt.Errorf("getting active boost deals: %w", err)
+	}
+
+	boostDeals = append(boostDeals, boostCompleteDeals...)
+	boostDeals = append(boostDeals, boostActiveDeals...)
```

Comment:
Seems like `boostDeals` size is known. Instantiate it with the right size after complete and active deal slices are determined? That should result in a more efficient allocation.